File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/GroundOverlayHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.GroundOverlay
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NGroundOverlay
import io.flutter.plugin.common.MethodChannel

internal interface GroundOverlayHandler : OverlayHandler {
    fun handleGroundOverlay(
        groundOverlay: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (groundOverlay as GroundOverlay).let { g ->
        when (method) {
            NGroundOverlay.boundsName -> setBounds(g, arg!!)
            NGroundOverlay.imageName -> setImage(g, arg!!)
            NGroundOverlay.alphaName -> setAlpha(g, arg!!)
            else -> result.notImplemented()
        }
    }

    fun setBounds(groundOverlay: GroundOverlay, rawBounds: Any)

    fun setImage(groundOverlay: GroundOverlay, rawNOverlayImage: Any)

    fun setAlpha(groundOverlay: GroundOverlay, rawAlpha: Any)
}

File: ios/Classes/controller/overlay/handler/GroundOverlayHandler.swift
import NMapsMap

internal protocol GroundOverlayHandler: OverlayHandler {
    func setBounds(_ groundOverlay: NMFGroundOverlay, rawBounds: Any)

    func setImage(_ groundOverlay: NMFGroundOverlay, rawNOverlayImage: Any)

    func setAlpha(_ groundOverlay: NMFGroundOverlay, rawAlpha: Any)
}

internal extension  GroundOverlayHandler {
    func handleGroundOverlay(groundOverlay: NMFGroundOverlay,
                             method: String,
                             args: Any?,
                             result: @escaping FlutterResult) {
        switch method {
        case NGroundOverlay.boundsName: setBounds(groundOverlay, rawBounds: args!)
        case NGroundOverlay.imageName: setImage(groundOverlay, rawNOverlayImage: args!)
        case NGroundOverlay.alphaName: setAlpha(groundOverlay, rawAlpha: args!)
        default: result(FlutterMethodNotImplemented)
        }
    }
}